### DISCUSSION TITLE
Adding IP addresses to devices without interface

### DISCUSSION THREAD

**Discussion by patrickgilsf** [netbox-community/netbox#8837] [Q&A] [answered]
Hi, I have a script that mass migrates our devices on network from Google Sheet to Netbox via rest API. It runs every day, and PATCHs any updates, and POSTs any new entries. The trickiest part here are IP addresses. The idea here is to add a device's IP, get the IP addresses' unique id, and then PATCH the device with the unique ID in the `primary_ip4` slot. However, my workflow doesn't (yet) involve _assigning_ the IP to the device, which I have discovered needs to be done via an extra step....so I get the following:
`primary_ip4: [
      'The specified IP address (10.14.24.11/32) is not assigned to this device.'
    ]`

In a perfect world, adding the IP address' unique id number as the `primary_ip4` parameter in a PATCH request, would do the assigning. Not sure why, but seems a bit more convoluted of a process than that. 

So, I'm looking at a deeper workflow laid out in [this thread](https://github.com/netbox-community/netbox/discussions/8746). The caveat here is that I don't manage netbox, and interfaces are not a device category that we use, which is a trend that predates me. As a matter of fact, all of our individual IPs are listed in as /32, and the requirement of me isn't to attach the IP Prefix, Interface, DHCP server/pool, or anything like that. We don't have our whole IP infra mapped out in netbox, but we would like to know what device is being pinged when we ping a device, and use Netbox as a tool to look that up.  

So, what steps can I take to achieve this goal, or can I not do it? I want every devices' /32 address associated with the devices I update. 

Until then, I am adding the IP address as a _comment_ in the device's page, but that's not really ideal for us, for a number of reasons you can guess. 

Any help is appreciated.

**Comment by jcollie** [comment, marked as answer]
I think that you're going to need to add interfaces to the devices as the
device/interface concept is core to NetBox, especially as it's a core
concept of networking in general (IP addresses aren't assigned to devices,
they are always assigned to interfaces). What I usually do when I have a
device that I want to attach an IP to is just to create a generic
"Ethernet" interface on the device. You'll need to specify a type but
1000base-t is generic enough. Once you have an interface you can assign the
IP to the interface and then assign the primary IP to the device.

On Wed, Mar 9, 2022 at 2:42 PM Patrick Gilligan ***@***.***>
wrote:

> Hi, I have a script that mass migrates our devices on network from Google
> Sheet to Netbox via rest API. It runs every day, and PATCHs any updates,
> and POSTs any new entries. The trickiest part here are IP addresses. The
> idea here is to add a device's IP, get the IP addresses' unique id, and
> then PATCH the device with the unique ID in the primary_ip4 slot.
> However, my workflow doesn't (yet) involve *assigning* the IP to the
> device, which I have discovered needs to be done via an extra step....so I
> get the following:
> primary_ip4: [ 'The specified IP address (10.14.24.11/32) is not assigned
> to this device.' ]
>
> In a perfect world, adding the IP address' unique id number as the
> primary_ip4 parameter in a PATCH request, would do the assigning. Not
> sure why, but seems a bit more convoluted of a process than that.
>
> So, I'm looking at a deeper workflow laid out in this thread
> <https://github.com/netbox-community/netbox/discussions/8746>. The caveat
> here is that I don't manage netbox, and interfaces are not a device
> category that we use, which is a trend that predates me. As a matter of
> fact, all of our individual IPs are listed in as /32, and the requirement
> of me isn't to attach the IP Prefix, Interface, DHCP server/pool, or
> anything like that. We don't have our whole IP infra mapped out in netbox,
> but we would like to know what device is being pinged when we ping a
> device, and use Netbox as a tool to look that up.
>
> So, what steps can I take to achieve this goal, or can I not do it? I want
> every devices' /32 address associated with the devices I update.
>
> Until then, I am adding the IP address as a *comment* in the device's
> page, but that's not really ideal for us, for a number of reasons you can
> guess.
>
> Any help is appreciated.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/netbox-community/netbox/discussions/8837>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFUVNSC6LJE4LC3ZTTFD7TU7EEKNANCNFSM5QKVG2EQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Jeff Ollie
The majestik møøse is one of the mäni interesting furry animals in Sweden.

**Reply by patrickgilsf** [reply]
Thanks. I'm just going through and creating mirror interfaces with same hostname, should work out. I wish there was a way to add an option to a device that it only has one interface, and as such, device and interface are the same thing...but this isn't as bad as I thought. Thanks for clarifying.

**Reply by patrickgilsf** [reply]
So, I am getting successful responses when I go to the ipam/ip-addresses/{id}/ endpoint, and adding in the `assigned_object_id`. They look like this:
`{
  id: 97438,
  url: 'https://<my company's server address>/api/ipam/ip-addresses/97438/',
  family: { value: 4, label: 'IPv4' },
  address: '10.6.17.28/32',
  vrf: null,
  tenant: null,
  status: { value: 'reserved', label: 'Reserved' },
  role: null,
  assigned_object_type: null,
  assigned_object_id: 39786,
  assigned_object: null,
  nat_inside: null,
  nat_outside: null,
  dns_name: '',
  description: 'Scanned and marked as alive',
  tags: [],
  custom_fields: { Scanned: 'Alive' },
  created: '2019-05-04',
  last_updated: '2022-03-10T02:26:12.193007Z'
}`

However, I try to add the `primary_ip4` to the devive with a PATCH request, and I still get `      'The specified IP address (172.30.24.141/32) is not assigned to this device.'`
Nor do I see any ip data populated when I look at a device or its associated interface. 


So, what is the exact method for taking a device and interface that are assigned to each other, and assigning this combo to an IP address? Because I thought it was to request PATCH `/ipam/ip-addresses/{id}/`, and provide the ipv4 address with mask (as above), and add the id for the object into the `assigned_object_id` field (also, as above). But no luck.

Thanks

**Reply by patrickgilsf** [reply]
I should add that I also just tried inputting the device interface id, instead of the device id, into `assigned_object_id`, and no luck. I'm thinking I might have the wrong parameter, but I'm not sure.

**Reply by markkuleinio** [reply]
You definitely need to assign the address to the interface ID (not device), and you are missing the `assigned_object_type`, which is `dcim.interface` IIRC.

**Reply by candlerb** [reply]
> What I usually do when I have a device that I want to attach an IP to is just to create a generic "Ethernet" interface on the device.

Even more generic is to select "Virtual"; then you're not making any claim about any physical port or connectivity.

But unfortunately you still do need to create Device, then Interface attached to Device, then IP address attached to Interface, and then update the Device to set the primary_ip4.

**Reply by patrickgilsf** [reply]
Thanks everyone, my issue, after realizing that I had been adding the device's id, not the interface id, was indeed assigning the string 'dcim.interface' to `assigned_object_type`. Oddly enough, neither of those properties have the red _required_ asterix in the docs. Code runs cleanly now.

**Comment by jcollie** [comment]
They aren't marked as required because you don't have to assign an IP to an
interface. However, when you do want to assign an IP address, you need to
specify both because IP addresses can be assigned to things other than
device interfaces (like VM interfaces).

On Thu, Mar 10, 2022 at 10:42 AM Patrick Gilligan ***@***.***>
wrote:

> Thanks everyone, my issue, after realizing that I had been adding the
> device's id, not the interface id, was indeed assigning the string
> 'dcim.interface' to assigned_object_type. Oddly enough, neither of those
> properties have the red *required* asterix in the docs. Code runs cleanly
> now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/netbox-community/netbox/discussions/8837#discussioncomment-2333089>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFUVNWTFMZ67P3KSTSUCIDU7IQ67ANCNFSM5QKVG2EQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 
Jeff Ollie
The majestik møøse is one of the mäni interesting furry animals in Sweden.

**Reply by patrickgilsf** [reply]
Thanks for the clarity on that.
